### PR TITLE
AM-2999 Uploading Certificates for mTLS

### PR DESCRIPTION
**Pull request by lgw-gravitee**
AM-2999

image

image

image

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/CertificatesResource.java [hunk 1]
```diff
@@ -257,6 +259,15 @@ private CertificateEntity filterCertificateInfos(Certificate certificate, int ce
...
         }
+        if (certificate.getConfiguration() != null) {
+            final JsonObject config = JsonObject.mapFrom(Json.decodeValue(certificate.getConfiguration(), HashMap.class));
+            if (config != null && config.getJsonArray("use") != null) {
+                filteredCertificate.setUsage(config.getJsonArray("use").stream().map(Object::toString).toList());
+            } else {
+                filteredCertificate.setUsage(List.of());
+            }
+        }
+
```
Comment (leleueri) [inline gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/CertificatesResource.java]: **Suggestion/Question**: Wondering if the logic to extract the _Use_ values should be moved into the **certificate-api** module into a Configuration Utility class :thinking:

File: gravitee-am-ui/src/app/domain/settings/certificates/certificates.component.ts [hunk 2]
```diff
@@ -139,7 +157,10 @@ export class DomainSettingsCertificatesComponent implements OnInit {
 
   rotateCertificate() {
     this.dialogService
-      .confirm('Rotate key', 'Are you sure you want to generate a new system certificate ?')
+      .confirm(
+        'Rotate key',
+        'Are you sure you want to generate a new system certificate?\nThis will only effect the latest System Certificate created.',
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/settings/certificates/certificates.component.ts:162]:
```suggestion
        'Are you sure you want to generate a new system certificate?\nThis will only affect the latest System Certificate created.',
```
Your change is consistent with figma design, but 'effect' looks strange to me :thinking:

**Review by leleueri — COMMENTED** [review]
Question: Wondering if we should prevent the mTLS usage for "system" certificate. As they are self signed certificates generated by AM, it make no sens to mark them as mTLS. It may be better UX to reject creation/update i this case. wdyt ?

**Comment by lgw-gravitee** [comment]
> **Question**: Wondering if we should prevent the mTLS usage for "system" certificate. As they are self signed certificates generated by AM, it make no sens to mark them as mTLS. It may be better UX to reject creation/update i this case. wdyt ?

Yes. It should be a part of next story https://gravitee.atlassian.net/browse/AM-2998 where we will use uploaded certs. Here, we can

File: gravitee-am-certificate/gravitee-am-certificate-api/src/main/java/io/gravitee/am/certificate/api/ConfigurationCertUtils.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public static List<String> extractUsageFromCertConfiguration(String configuration) {
+        if (configuration == null) {
+            return List.of();
+        }
+        try {
+            Map<String, Object> cfg = JSONObjectUtils.parse(configuration);
+            List<String> uses = JSONObjectUtils.getStringList(cfg, "use");
+            return uses == null ? List.of() : uses;
+        } catch (ParseException e) {
+            log.debug("Problem at parsing certificate configuration", e);
```
Comment (leleueri) [inline gravitee-am-certificate/gravitee-am-certificate-api/src/main/java/io/gravitee/am/certificate/api/ConfigurationCertUtils.java:42]: **question**: should it be a WARN ? I'm OK to return an empty list but having a parsing exception here is maybe something we whould be worried as it should no happen. wdyt ?

**Review by leleueri — APPROVED** [review]
LGTM, I let you decide if the log level need to be changed.

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/test/java/io/gravitee/am/management/handlers/management/api/resources/CertificatesResourceTest.java [hunk 4]
```diff
@@ -68,7 +70,64 @@ public void shouldGetCertificates() {
...
+
+
+        doReturn(Maybe.just(mockDomain)).when(domainService).findById(domainId);
+        doReturn(Flowable.just(mockCertificate, mockCertificate2, mockCertificate3, mockCertificate4)).when(certificateService).findByDomain(domainId);
+        doReturn(Flowable.empty()).when(applicationService).findByCertificate(anyString());
+
+        final Response response = target("domains").path(domainId).path("certificates").request().get();
+        assertEquals(HttpStatusCode.OK_200, response.getStatus());
+
+        final List<LinkedHashMap> responseEntity = readEntity(response, List.class);
```
Comment (podlesrafal) [inline gravitee-am-management-api/gravitee-am-management-api-rest/src/test/java/io/gravitee/am/management/handlers/management/api/resources/CertificatesResourceTest.java:114]: There is readListEntity method. You can use it like this:  readListEntity(reponse, CertificateEntity.class)

File: gravitee-am-model/src/main/java/io/gravitee/am/model/Certificate.java [hunk 5]
```diff
@@ -18,6 +18,7 @@
 import io.swagger.v3.oas.annotations.media.Schema;
 
 import java.util.Date;
+import java.util.List;
```
Comment (podlesrafal) [inline gravitee-am-model/src/main/java/io/gravitee/am/model/Certificate.java:21]: remove

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/CertificateServiceImpl.java [hunk 6]
```diff
@@ -18,6 +18,7 @@
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.Lists;
```
Comment (podlesrafal) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/CertificateServiceImpl.java:21]: remove unused imports

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/CertificateServiceImpl.java [hunk 7]
```diff
@@ -120,6 +126,7 @@ public class CertificateServiceImpl implements CertificateService {
     private static final String STORE_PASS = "storepass";
     private static final String ALIAS = "alias";
     private static final String NAME = "name";
+    private static final String USAGE = "use";
```
Comment (podlesrafal) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/CertificateServiceImpl.java:129]: remove

File: gravitee-am-certificate/gravitee-am-certificate-api/src/main/java/io/gravitee/am/certificate/api/ConfigurationCertUtils.java (hunk 3 above)
Comment (lgw-gravitee) [reply gravitee-am-certificate/gravitee-am-certificate-api/src/main/java/io/gravitee/am/certificate/api/ConfigurationCertUtils.java]: You're right, I've changed to warn.

**Pull request merged**